ta/remoteproc/remoteproc_core.c

```diff
@@ -0,0 +1,1102 @@
+ *                 /      +-------------+  0x3543A468
+ *                /       +-------------+
+ *               /        |   version   |  32-bit word, version of the format
+ *              /         +-------------+
+ * +-----------+          +-------------+
+ * |   Header  |          |  TLV size   |  32-bit word, size of the TLV
+ * +-----------+          +-------------+  (aligned on 64-bit), in bytes.
+ *              \         +-------------+
+ *               \        |  sign size  |  32-bit word, size of the signature
+ *                \       +-------------+  (aligned on 64-bit), in bytes.
+ *                 \      +-------------+
+ *                  \     | images size |  32-bit word, size of the images to
+ *                   -----+-------------+  load (aligned on 64-bit), in bytes.
+ *
+ *                        +-------------+  Information used to authenticate the
+ *                        |     TLV     |  images and boot the remote processor,
+ *                        |             |  stored in Type-Length-Value format.
+ *                        +-------------+  'Type' and 'Length' are 32-bit words.
+ *
+ *                        +-------------+
+ *                        | Signature   |   Signature of the header and the TLV.
+ *                        +-------------+
+ *
+ *                        +-------------+
+ *                        |   Firmware  |
+ *                        |    image 1  |
+ *                        +-------------+
+ *                               ...
+ *                        +-------------+
+ *                        |   Firmware  |
+ *                        |    image n  |
+ *                        +-------------+
+ */
+
+/* Firmware state */
+enum remoteproc_state {
+	REMOTEPROC_OFF = 0,
+	REMOTEPROC_LOADED,
+	REMOTEPROC_STARTED,
+};
+
+#define RPROC_HDR_MAGIC		0x3543A468 /* random value */
+#define HEADER_VERSION		1
+
+/* Supported signature algorithm */
+enum remoteproc_sign_type {
+	RPROC_RSASSA_PKCS1_v1_5_SHA256 = 1,
+	RPROC_ECDSA_SHA256 = 2,
+};
+
+enum remoteproc_img_type {
+	REMOTEPROC_ELF_TYPE = 1,
+	REMOTEPROC_INVALID_TYPE = 0xFF
+};
+
+/* remoteproc_tlv structure offsets */
+#define RPROC_TLV_LENGTH_OF	U(0x04)
+#define RPROC_TLV_VALUE_OF	U(0x08)
+
+/* TLV types */
+#define RPROC_TLV_SIGNTYPE	U(0x00000001)
+#define RPROC_TLV_HASHTYPE	U(0x00000002)
+#define RPROC_TLV_NUM_IMG	U(0x00000003)
+#define RPROC_TLV_IMGTYPE	U(0x00000004)
+#define RPROC_TLV_IMGSIZE	U(0x00000005)
+#define RPROC_TLV_HASHTABLE	U(0x00000010)
+#define RPROC_TLV_PKEYINFO	U(0x00000011)
+
+#define RPROC_PLAT_TLV_TYPE_MIN	U(0x00010000)
+#define RPROC_PLAT_TLV_TYPE_MAX	U(0x00020000)
+
+#define RPROC_TLV_SIGNTYPE_LGTH U(1)
+
+#define LE16_TO_CPU(x) ({		\
+		uint8_t *__x = x;	\
+		(__x[1] << 8) | __x[0]; \
+	})
+#define LE32_TO_CPU(x) ({					         \
+		uint8_t *__x = x;					 \
+		(__x[3] << 24) | (__x[2] << 16) | (__x[1] << 8) | __x[0]; \
+	})
+
+#define U64_ADD_TO_ALIGN(s) ({					   \
```

Comment:
I'm not sure what these macros are supposed to do. How about adding a comment and implementing them using `ROUNDUP()` and friends instead?

Reply:
done